infix/eval_test.go

```diff
@@ -0,0 +1,118 @@
+package infix
+
+import (
+	"fmt"
+	"testing"
+)
+
+var format = fmt.Sprintf
+
+func desc(n Node, res Int) string {
+	if n.Type() == NodeBinExpr {
+		a := n.(BinExpr)
+		return format("%s %s %s = %s",
+			a.Lhs, a.Rhs, a.Op, res)
+	} else if n.Type() == NodeUnaryExpr {
+		a := n.(UnaryExpr)
+		return format("%s%s", a.Op, a.Value)
+	}
+
+	a := n.(Int)
+	return format("%d", a)
+}
+
+func TestEval(t *testing.T) {
```

Comment:
On this case it is more easy to understand but on nash I had a real pain on understanding eval tests that are performed on nodes that are built by hand like this, it is much more hard to understand than something like:

```
expr: "0 & 0",
res: 0
}
```

But this would imply tests that covers the parsing too, together with eval. But they are still isolated on the sense that there is no source of non-determinism and avoid duplicating test coverage (you need to test the parsing too I think). So on nash I was pretty torn if it would not be better to always test on top of the parser/lexer instead of building the AST nodes by hand on the tests.

Another advantage are tests that are on top of the language specification, not the design of node types and the AST. But I'm not sure, just brainstorming ;-).